Wrangle benchmarks

kazimuth opened the pull request:
# Description of Changes

Rewrite benchmarks to allow direct comparison between:
- `spacetime_module`: Spacetime via a module
- `spacetime_raw`: Spacetime via the RelationalDB struct
- `sqlite`

And add some targeted serialization benchmarks.

kazimuth committed:
Major benchmarks refactoring

kazimuth commented:
benchmarks please

kazimuth committed:
Fix compile after cherry-pick

kazimuth committed:
More cleanup

kazimuth committed:
Add some serialization benchmarks

kazimuth committed:
Refactor

kazimuth committed:
Provoke github actions

kazimuth committed:
Benchmark action needs clippy

kulakowski commented on crates/bench/src/lib.rs:
```diff
@@ -1,13 +1,125 @@
+        });
+    }
+
+    fn sanity<DB: BenchDatabase, T: BenchTable + RandomTable>(
```
I'd slightly prefer `check_basic_invariants` or something that's more specific.

kulakowski commented on crates/bench/src/lib.rs:
```diff
@@ -1,13 +1,125 @@
+        let table_id = db.create_table::<T>(table_style)?;
+        assert_eq!(db.count_table(&table_id)?, 0, "tables should begin empty");
+
+        let count = 37;
```
Comment or constant from elsewhere if possible.

kulakowski commented on modules/benchmarks/src/lib.rs:
```diff
@@ -1,183 +1,359 @@
+    x: u64,
+    y: u64,
+}
+// ---------- / SYNCED CODE ----------
```
Maybe `END SYNCED CODE` instead of html-style `/ SYNCED CODE`

kulakowski commented on modules/benchmarks/src/lib.rs:
```diff
@@ -1,183 +1,359 @@
+#[spacetimedb(reducer)]
+pub fn empty() {}
+
+// ---------- insert ----------
```
One kind of documentation would be super useful for this, and that's the framework that you used to derive all of these. Suppose we add some new operation, I'd love to then be able to methodically add the "correct" set of new things here.

kulakowski commented on crates/bench/src/sqlite.rs:
```diff
@@ -1,13 +1,36 @@
-    fn build(prefill: bool, fsync: bool) -> ResultBench<Self>
+pub struct SQLite {
+    db: Connection,
+    _temp_dir: TempDir,
```
The finickiness of the TempDir drop stuff makes me want a comment here, we've had enough bugs in tests and CI about it that I don't want this refactored away to a local or whatever.

kazimuth replied:
In fact I think refactoring it to a local could cause the tempdir to be dropped, which might break things... then again, the earlier benchmarks version did that and the sqlite benches still worked. So idk.

kulakowski commented on crates/bench/src/sqlite.rs:
```diff
@@ -16,81 +39,200 @@ impl BuildDb for Connection {
 
-        if prefill {
-            prefill_data(&mut db, Runs::Small)?;
+    fn create_table<T: BenchTable>(&mut self, table_style: crate::schemas::TableStyle) -> ResultBench<Self::TableId> {
```
Mario or someone might be a better reviewer of the sqlite stuff here than me.

kulakowski commented on crates/bench/src/spacetime_raw.rs:
```diff
@@ -0,0 +1,158 @@
+    }
+
+    type PreparedFilter = PreparedFilter;
+    #[inline(never)]
```
I picked this `#[inline(never)]` one arbitrarily out of all of them to comment: you should describe where and why you do this in some top level place if possible!

kulakowski commented on crates/bench/src/spacetime_module.rs:
```diff
@@ -0,0 +1,232 @@
+
+pub struct SpacetimeModule {
+    runtime: Runtime,
+    // it's necessary for this to be dropped BEFORE the runtime.
```
I personally prefer this (an explicit Drop impl) to e.g. relying on a comment about relying on the drop ordering based on field order. Maybe Mazdak or Noa have a stronger style opinion about it, I'd be curious.

kulakowski commented on crates/bench/src/schemas.rs:
```diff
@@ -0,0 +1,332 @@
+}
+// ---------- / SYNCED CODE ----------
+
+/// Note: the first field will be used as the primary key, when using
```
I can't understand how this comment relates to this trait. The first field of what?

kazimuth replied:
Of the product type / value conversion of the underlying struct, I'll move it.

kulakowski reviewed (approved):
Awesome. Inline is mostly minor comments about comments. I think this PR can land whenever, as far as I'm concerned, but I want two things.

One is the framework or schema or strategy or whatever you want to call it, by which you derived the set of benchmarks to write out. If Joshua (say) adds an Update operation that isn't just a delete+insert (say), I'd love for him to be able to read that framework and just slot in exactly the right set of benchmarks. I'd like this to be part of this PR.

Two, if there's a ways to go, I'd like other people to be able to see either in the code or in a ticket how far there is to go until this work is complete. I'm happy if there are more PRs following this one until you get there.

kazimuth commented on crates/bench/benches/special.rs:
```diff
@@ -0,0 +1,106 @@
+use spacetimedb_testing::modules::start_runtime;
+
+fn criterion_benchmark(c: &mut Criterion) {
+    /*
```
These benchmarks won't change frequently and are mostly only interesting to people writing benchmarks, but I used them to make a decision at one point so left them in. I'll add a comment.

kazimuth commented:
@kulakowski, I'll address your comments and then I think this is ready for merge.

> One is the framework or schema or strategy or whatever you want to call it, by which you derived the set of benchmarks to write out. If Joshua (say) adds an Update operation that isn't just a delete+insert (say), I'd love for him to be able to read that framework and just slot in exactly the right set of benchmarks. I'd like this to be part of this PR.

Good point, I'll document this.

> Two, if there's a ways to go, I'd like other people to be able to see either in the code or in a ticket how far there is to go until this work is complete. I'm happy if there are more PRs following this one until you get there.

Yeah, I wasn't really sure where I was going while working on this which meant I had a hard time communicating what I was doing. I have some more targeted benchmark ideas which I'll stick in a followup PR.

I do have one remaining concern, which is benchmark time. Looking at the tests on this PR, it seems that running benchmarks on GitHub actions currently takes about 80 minutes. I could strip out some / all of the sqlite benchmarks since they won't change frequently. We could also disable the in-memory benchmarks on GitHub actions, which would get rid of one axis of this giant hypercube. I think both of those provide useful information, though.

kazimuth commented on crates/core/src/host/wasm_common/module_host_actor.rs:
```diff
@@ -276,6 +276,21 @@ impl<T: WasmModule> Module for WasmModuleHostActor<T> {
+            let tables = db.get_all_tables(tx)?;
+            for table in tables {
+                if table.table_name != table_name {
+                    continue;
```
I wasn't sure if there could be multiple tables with the same name here, can there be?

kazimuth commented:
Oh, also, should I rename "Person" to "IntStringInt" and "Location" to "IntIntInt" or something similar?

kazimuth commented:
I have a path to fix benchmark times -- it's because everything is being run twice, including the sqlite benchmarks. I'll mess with our Github Actions benchmark script to fix that.

kazimuth committed:
Simplify bench database interface

kazimuth committed:
Address review comments

mamcx commented on crates/bench/src/sqlite.rs:
```diff
@@ -16,81 +47,238 @@ impl BuildDb for Connection {
             "PRAGMA journal_mode = WAL; PRAGMA synchronous = off;"
```
Is possible to impl the TODO here?

kazimuth replied:
Done!

kazimuth committed:
Better comments, s/table_style/index_strategy

kazimuth commented on crates/bench/src/sqlite.rs:
```diff
@@ -16,81 +47,238 @@ impl BuildDb for Connection {
+    }
+}
+
+const BEGIN_TRANSACTION: &str = "BEGIN DEFERRED";
```
The rusqlite transaction API actually just invokes these statements, but it doesn't cache them, which ends up penalizing performance significantly. So we do the same thing but cache it.

kazimuth committed:
Fix filesystem issues (names too long)

kazimuth committed:
Minor edits

mamcx reviewed (approved):
LGTM

gefjon committed:
Merge branch 'master' into kazimuth/benchwrangle

gefjon committed:
Replace now-removed `Table::delete` calls with `unimplemented!`

These calls unconditionally panicked before anyway;
now they just panic in a way that Clippy doesn't object to.

gefjon committed:
Merge branch 'master' into kazimuth/benchwrangle
